hack/pcr-compare/main.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package main
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+
+	"github.com/edgelesssys/constellation/v2/internal/attestation/measurements"
+	"github.com/fatih/color"
+)
+
+// Based on image/measured-boot/precalculate_pcr_<4,9,12>.sh & pcr-stable.json
+// This holds hex-encoded PCR values.
+type expectedMeasurementsFile struct {
+	Measurements expectedMeasurements `json:"measurements"`
+}
+
+type expectedMeasurements struct {
+	PCR4  string `json:"4"`
+	PCR8  string `json:"8"`
+	PCR9  string `json:"9"`
+	PCR11 string `json:"11"`
+	PCR12 string `json:"12"`
+	PCR13 string `json:"13"`
+}
+
+func main() {
+	if len(os.Args) < 3 {
+		if len(os.Args) == 0 {
+			fmt.Println("Usage:", "pcr-compare", "<expected-measurements> <actual-measurements>")
+		} else {
+			fmt.Println("Usage:", os.Args[0], "<expected-measurements> <actual-measurements>")
+		}
+		fmt.Println("<expected-measurements> is supposed to be a JSON file from the 'Build OS image' pipeline.")
+		fmt.Println("<actual-measurements> in supposed to be a JSON file with metadata from the PCR reader which is supposed to be verified.")
+		os.Exit(1)
+	}
+
+	parsedExpectedMeasurements, err := parseExpectedMeasurements(os.Args[1])
+	if err != nil {
+		panic(err)
+	}
+
+	parsedActualMeasurements, err := parseActualMeasurements(os.Args[2])
+	if err != nil {
+		panic(err)
+	}
+
+	mapExpectedMeasurements := createMapFromStrings(parsedExpectedMeasurements)
+	mapActualMeasurements := convertBytesMeasurementsToHexMeasurements(parsedActualMeasurements)
+
+	foundMismatch := compareMeasurements(mapExpectedMeasurements, mapActualMeasurements)
```

Comment:
This code seems unnecessarily complex.

Why not just unmarshal the measurement files directly to `measurements.M`, delete the keys we don't want (or copy just the keys we want), and then call `expectedMeasurements.EqualTo(actualMeasurements)` ?

We could even go with removing this tool completetly, and just use [`yq compare`](https://mikefarah.gitbook.io/yq/v/v3.x/commands/compare) to handle the task for us

Reply:
To be honest I would keep it in Go since it's cleaner than `yq compare`, has nicer output and would avoid more Bash hackery in the actions. Also the final consumer is going to be Go code using the same structs so it would make sense to have some common ground here.

But I'll try to strip down the code, I originally wrote it before we had `measurements.M`and `EqualTo`.

Reply:
Okay, I simplified this quite a bit. Now both inputs use the same functions for everything, equality is checked with `EqualTo` but I do the same output since I hope that's nice to look at for debugging.

Check if you think it's fine that way.

Reply:
Okay, with cooler output it becomes slightly more complicated, though the actual check is still done by `EqualTo` so it should fail and abort the pipeline even if the more complex output logic has some mistakes.

On the other hand though, `yq compare` is deprecated and does not exist anymore:
https://mikefarah.gitbook.io/yq/upgrading-from-v3#comparing-yaml-files

Of course we could diff but I don't know, I still like the Go code more.

Reply:
Looks a lot better now, thanks.

IMO the extra comparison is not that important, since we already print the files multiple times during the workflow, but if you think it adds value we can keep it.

Reply:
Alright.

 If this thing makes too much trouble in the future we can just remove it and replace it I guess. But I like it for fast debugging (though it looks cooler locally to be honest since Github Actions strips the colors).